nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/ConsumeSlack.java

```diff
@@ -0,0 +1,234 @@
+    .required(true)
+    .identifiesControllerService(SlackConnectionService.class)
+    .build();
+
+  static final PropertyDescriptor MESSAGE_TYPES = new PropertyDescriptor
+    .Builder()
+    .name("message-types")
+    .displayName("Message types")
+    .description("Message types to listen to. It will filter messages with the given types or " +
+      "listen to every type if empty. It is a coma separated list like: message,file_shared")
+    .required(false)
+    .addValidator(Validator.VALID)
+    .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+    .build();
+
+  public static final PropertyDescriptor MAX_MESSAGE_QUEUE_SIZE = new PropertyDescriptor.Builder()
+    .name("Max Size of Message Queue")
+    .description("The maximum size of the internal queue used to buffer messages being transferred from the underlying channel to the processor. " +
+      "Setting this value higher allows more messages to be buffered in memory during surges of incoming messages, but increases the total " +
+      "memory used by the processor.")
+    .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+    .defaultValue("10000")
+    .required(true)
+    .build();
+
+  static final Relationship SUCCESS_RELATIONSHIP = new Relationship.Builder()
+    .name("success")
+    .description("Incoming messages that matches the given types")
+    .build();
+
+  private static final String JSON_OBJECT_TYPE_KEY = "type";
+
+  private static final List<PropertyDescriptor> DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(
+    SLACK_CONNECTION_SERVICE, MESSAGE_TYPES, MAX_MESSAGE_QUEUE_SIZE));
+
+  private static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+    SUCCESS_RELATIONSHIP)));
+
+  @Override
+  public Set<Relationship> getRelationships() {
+    return RELATIONSHIPS;
+  }
+
+  @Override
+  public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return DESCRIPTORS;
+  }
+
+  @OnScheduled
+  public void onScheduled(final ProcessContext context) {
+    String messageTypes = context.getProperty(MESSAGE_TYPES).evaluateAttributeExpressions().getValue();
+    messageTypes = messageTypes == null ? "" : messageTypes;
+    matchAny = messageTypes.isEmpty();
+    matchingTypes = Arrays.asList(messageTypes.split(","));
+    messageQueue = new LinkedBlockingQueue<>(context.getProperty(MAX_MESSAGE_QUEUE_SIZE).asInteger());
+  }
+
+  @OnUnscheduled
+  @OnShutdown
+  public void onUnscheduled() {
+    if (isProcessorRegisteredToService()) {
+      deregister();
+    }
+  }
+
+  private void deregister() {
+    slackConnectionService.deregisterProcessor(this);
+  }
+
+  @Override
+  public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+    if (!isProcessorRegisteredToService()) {
+      registerProcessorToService(context);
+    }
+
+    SlackMessage slackMessage = messageQueue.poll();
```

Comment:
Currently, a onTrigger process only one slackMessage. I'd suggest adding `@SupportsBatching` annotation. Which adds following `Run Duration` slider so that user can configure performance characteristics.
![image](https://user-images.githubusercontent.com/1107620/55127113-65499500-5152-11e9-81bd-2901a23b9ec3.png)

However, `Run Duration` may not work well with these source type processor, because the annotation makes NiFi scheduler keeps running onTrigger for the configured duration. If there's no messages in the `messageQueue`, it wastes CPU times.

Alternatively, we can implement a while loop here, polling messages until it gets null. If we go with that approach, I also suggest adding something like `Batch Size`.